docs/proposals/201909_bucketfile.md

```diff
@@ -0,0 +1,329 @@
+---
+title: BucketFile for reducing memory usage 
+type: proposal
+menu: proposals
+status: 
+owner: zhulongcheng
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/1471 (main)
+* https://github.com/thanos-io/thanos/issues/448 (store gateway OOM)
+
+## Summary
+
+This document describes the motivation and design of the `BucketFile` of the store component. 
+
+Our goals here are:
+
+- A) Reduce memory usage
+- B) Reuse chunks fetched from object storage
+
+## Motivation
+
+Currently, the store gateway loads all [index.cache.json](https://github.com/thanos-io/thanos/blob/865d5ec710b5651ead08a7dc73864a6f5dbeeaa9/pkg/block/index.go#L40)  files into memory on startup, which easily hits OOM situations. On the other hand, some indexes are rarely or never queried, so we don't have to keep these indexes in memory.
+
+The store gateway fetches chunks from object storage for each query request. but fetched chunks are not reused for next requests.
+
+## Proposal
+
+Add a helper struct to wrapper remote object storage file. We can temporarily name it `BucketFile`.
+
+`BucketFile`   looks like `ByteSlice` in  prometheus/tsdb, but with some download-with-cache functions. It fetches part or all of file content from object storage, writes content to local file, and records data have been fetched by a bitmap.
+
+`BucketFile` always returns a slice of the mmaped file. This is to reduce coping file data from kernel space to user space. 
+
+### BucketFile in a block
+
+
+
+![BucketBlock](./img/bucketfile_in_block.jpg)
+
+The above sequence diagram shows how the BucketFile works with IndexReader and ChunksReader in a block:
+
+(BucketIndexFile and BucketChunksFile are represented by `BucketFile` )
+
+1. BucketStore calls `blockSeries`  method to retrieve series from a block.
```

Comment:
It might be me but I feel like all of this description might be a bit too verbose. We should focus on how `BucketFile` works

Reply:
this helps to understand how `BucketFile` works with other components.

will remove if no need.

Reply:
Essentially what is here is just coding it straight away but in human language. That's hard to review (:  It's similar if we would just start coding without design. We might need to agree on high level things first.

Reply:
Ok, will simplify this.